libdnf5/repo/repo_downloader.cpp

```diff
@@ -493,6 +513,28 @@ void RepoDownloader::common_handle_setup(LibrepoHandle & h) {
     if (optional_metadata.extract(libdnf5::METADATA_TYPE_UPDATEINFO)) {
         dlist.push_back(MD_FILENAME_UPDATEINFO);
     }
+    if (get_optional_metadata().extract(libdnf5::METADATA_TYPE_APPSTREAM)) {
+        // ideally, the repomd.xml file should be read and every type matching is_appstream_metadata_type()
+        // would be added from it, but the content is not known at this point and when it is known, then
+        // it's too late, thus declare some "expected" types to be downloaded here
+        dlist.push_back("appstream");
+        dlist.push_back("appstream-icons");
+        dlist.push_back("appstream-icons-48x48");
+        dlist.push_back("appstream-icons-48x48@2");
+        dlist.push_back("appstream-icons-64x64");
+        dlist.push_back("appstream-icons-64x64@2");
+        dlist.push_back("appstream-icons-128x128");
+        dlist.push_back("appstream-icons-128x128@2");
+        // consult the prefixes with the is_appstream_metadata_type()
+        dlist.push_back("appdata");
+        dlist.push_back("appdata-icons");
+        dlist.push_back("appdata-icons-48x48");
+        dlist.push_back("appdata-icons-48x48@2");
+        dlist.push_back("appdata-icons-64x64");
+        dlist.push_back("appdata-icons-64x64@2");
+        dlist.push_back("appdata-icons-128x128");
+        dlist.push_back("appdata-icons-128x128@2");
```

Comment:
So we want to always have first just the repomd present to parse appstream data for that. We have currently these scenarios when downloading metadata:

1. No previous metadata => downloading everything once
2. Previous metadata present but old => first download only repomd to check it's old, then download everything (including repomd again)
3. Previous metadata present and valid => only download repomd to check it's valid

It would mean that we just need to alter the 1. scenario to have the additional download of repomd before downloading everything.

I think it should be manageable. Just it would need some refactoring, the repomd-only downloading is currently done in `RepoDownloader::is_repomd_in_sync()`. In our case we want to reuse the `LrYumRepo` object from there for our appstream parsing.

Reply:
Hmm, that could work, yes.

Is it acceptable to update the "global" `LRO_YUMDLIST` in the `is_repomd_in_sync()`? The function would do more than just checking whether the repomd is up-to-date afterwards.

I already lost my code where I've been playing with these things, but it should not be that hard to write it again.

Reply:
If metadata are correctly downloaded to cache, they should not be downloaded again during the next run. They are loaded from cache then, the paths should be stored in `metadata_paths`. So maybe it could be also a problem with that. Would like to debug it, but unfortunately don't have capacity for that in upcoming days...

Reply:
OK, let's keep things as they are. I will create a follow-up issue to try optimizing those hardcoded names.